### DISCUSSION TITLE
PetVision_37

### DISCUSSION THREAD

**Discussion by ryujin-akm** [mrdbourke/pytorch-deep-learning#180] [Show and tell]
PetVision_37 -> [click_this_link_for_the_webpage](https://huggingface.co/spaces/ryu-akm/PetVision_37)

How is this project different from the FoodVision Big and mini ?

The first difference is the file structure. In the Food101 dataset the images came pre arranged in the classes folder.
(Ex -> pizza -> image1, image2   , steak -> image3, image4)

In Pet dataset ([Oxford-IIIT Pet Dataset](https://www.robots.ox.ac.uk/~vgg/data/pets/)) the images are not arranged in a similar fashion so error comes when using the dataloader. For this reason the first thing I did was separate the images and move them in the similar fashion as expected by the Imagefolder.
 
Second, this dataset is quite small(200 images per class) compared to the Food101 dataset (1000 images per class).
Still I was able to achieve quite high accuracy (90%), considering the original paper's accuracy of 59% we were able to beat it with a significant margin. This goes to show how far Computer Vision techiniques have come.

**Comment by mrdbourke** [comment]
Hey @ryujin-akm!

Massive effort creating your own deployed model!!!

That's a huge step towards practising what you've learned.

I'm trying out your application and the example image works but it seems to be stuck loading on custom images 🤔 

Might be an error on HuggingFace's behalf.

Have you tried it with your own custom images, does it work?

<img width="1624" alt="Screenshot 2022-11-17 at 10 40 57 am" src="https://user-images.githubusercontent.com/16750345/202325790-c7d3f6be-ba31-40c1-8122-49b1e35a06d8.png">

**Reply by shisirkha** [reply]
Sir hehe you got a typo in there it will be word practicing not practising.

**Comment by mrdbourke** [comment]
Oh wait it worked!

Looks like the model thinks my dog Bella (my labrador/border collie) is a Shiba Inu! Hahaha I love it!

Perhaps labrador isn't in the class names list?

<img width="1624" alt="Screenshot 2022-11-17 at 10 44 36 am" src="https://user-images.githubusercontent.com/16750345/202326309-8b1674ad-9eed-4296-b091-1cdc11ecbd8e.png">

**Reply by ryujin-akm** [reply]
Yep labrador/bordor collie are not in the list. I think can go for something you suggested in the videos, like if the class is not there, then it would say "class not found or somthing", I tired that experiment with small dataset, so if accuracy was less than 50% it would say "accuracy less than 50 class not found", but this dosen't feel optimum choice.

I tried with a bunch of custom images, one interesting thing was unable to differentiate between similar breeds despite high accuracy (example -> birman and ragdoll, cat breed).
![image](https://user-images.githubusercontent.com/109247925/202409493-79183cdb-3e16-4971-ac88-81830059bd30.png)

![image](https://user-images.githubusercontent.com/109247925/202409158-69b1e7e2-681d-49c0-9197-18f3b7029b0d.png)


Another point is background is also an issue. I am sharing the pics.
![image](https://user-images.githubusercontent.com/109247925/202408650-2060c312-887a-435e-b128-153c3cce24b8.png)


I think it's due to the test error being around 1.02. Tried a bunch of ways but couldn't reduce it so I guess just wanted to see the model deployed first 😅.

I will increase the test set size and I think that should help. My first thought was how less data we can go for ? So I only took 20% for test, the test images only being 100 that's just 20 images. 

Anycase loved to hear that the model is working for you as well (although not accurately), And would love to hear your thoughts on the test error issue.

**Reply by ryujin-akm** [reply]
@mrdbourke

**Reply by mrdbourke** [reply]
Hi @ryujin-akm,

In this case it looks like increasing the training size would help more.

It makes sense for the model to get confused on similar classes though.

For example, one of my food vision models gets confused on Cucumbers/Zucchini's (they look very similar).

Things you could try: 
* Increase training data (find more images of the most confusing classes on the internet)
* Try a bigger model/training for longer (potentially there are other transfer learning models out there you could try)
* Create another model for "no animal/animal" and run your image through that model as well, if there's no animal found, the model outputs "no pet found please try another image"

**Reply by Apoorv1401** [reply]
Hi @mrdbourke,

Yeah I will try the adjustments. Yep more examples should surely help and I will try a little bit a fine-tuning as well.

Happy New Year !!

**Comment by teacherc** [comment]
@ryujin-akm Congrats on this awesome project!

**Reply by Apoorv1401** [reply]
Hi @teacherc,

Thank you so much. You are being kind. 
Happy New Year!!

**Comment by ryujin-akm** [comment]
That's my other account so don't worry about it being imposter :grin:
